src/material_library.cpp

```diff
@@ -0,0 +1,342 @@
+  std::pair<pyne::matname_set::iterator, bool> mat_insert;
+  if (mat.metadata.isMember("name")) {
+    mat_name = mat.metadata["name"].asString();
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+  } else {
+    // form a mat name as 'mX'
+    int mat_number = name_order.size() +1;
+    mat_name = "m" + std::to_string(mat_number);
+    mat.metadata["name"] = mat_name;
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+
+    // if 'mX' name is already in the Library rename the material
+    while (!mat_insert.second) {
+      mat_number += name_order.size();
+      mat_name = "m" + std::to_string(mat_number);
+      mat.metadata["name"] = mat_name;
+      mat_insert = matlist.insert(mat.metadata["name"].asString());
+    }
+  }
+  if (mat_insert.second) {
+    append_to_nuclist(mat);
+    material_library[mat_name] = new Material(mat);
+    material_library[mat_name]->metadata["mat_number"] = int(name_order.size() +1);
+    
+    name_order.push_back(mat_name);
+  }
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat, char* mat_name) {
+  std::string material_name(mat_name);
+  add_material(mat, material_name);
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat,
+                                         const std::string& mat_name) {
+  mat.metadata["name"] = mat_name;
+  add_material(mat);
+}
+
+void pyne::MaterialLibrary::del_material(pyne::Material mat) {
+  if (mat.metadata.isMember("name")) {
+    std::string mat_name = mat.metadata["name"].asString();
+    del_material(mat_name);
+  }
+}
+
+void pyne::MaterialLibrary::del_material(const std::string& mat_name) {
+  material_library.erase(mat_name);
+  matlist.erase(mat_name);
+  for ( auto name = name_order.begin(); name != name_order.end(); name++) {
+    if (*name == mat_name) {
+      name_order.erase(name);
+      return;
+    }
+  }
+}
+
+pyne::Material pyne::MaterialLibrary::get_material(
+    const std::string& mat_name) {
+  auto it = material_library.find(mat_name);
+  if (it != material_library.end()) {
+    return *(it->second);
+  } else {
+    return pyne::Material();
+  }
+}
+
+pyne::Material* pyne::MaterialLibrary::get_element(const std::string& mat_name) {
+  auto it = material_library.find(mat_name);
+  if (it != material_library.end()) {
+    return it->second;
+  } else {
+    return new pyne::Material();
+  }
+}
+
+pyne::Material* pyne::MaterialLibrary::get_element_by_indice(int num) {
+  if (num < name_order.size()) {
+    return material_library[name_order[num]];
+  } else {
+    return new pyne::Material();
+  }
+}
+
+void pyne::MaterialLibrary::replace(int num, pyne::Material mat) {
+
+  if (!mat.metadata.isMember("name")) {
+    mat.metadata["name"] = name_order[num];
+  } else if (mat.metadata["name"] != name_order[num]) {
+    material_library.erase( name_order[num]);
+    name_order[num] = mat.metadata["name"].asString();
+  }
+  *material_library[name_order[num]] = name_order[num];
```

Comment:
This should be storing `mat` in the appropriate way...

(and make sure your tests are setup correctly :) )